File: token-metadata/js/src/transactions/CreateMasterEdition.ts
import { Borsh } from '@metaplex/utils';
import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
import {
  PublicKey,
  SystemProgram,
  SYSVAR_RENT_PUBKEY,
  TransactionCtorFields,
  TransactionInstruction,
} from '@solana/web3.js';
import BN from 'bn.js';
import { Transaction } from '../../../Transaction';
import { MetadataProgram } from '../MetadataProgram';

export class CreateMasterEditionArgs extends Borsh.Data<{ maxSupply: BN | null }> {
  static readonly SCHEMA = this.struct([
    ['instruction', 'u8'],
    ['maxSupply', { kind: 'option', type: 'u64' }],
  ]);

  instruction = 10;
  maxSupply: BN | null;
}

type CreateMasterEditionParams = {
  edition: PublicKey;
  metadata: PublicKey;
  updateAuthority: PublicKey;
  mint: PublicKey;
  mintAuthority: PublicKey;
  maxSupply?: BN;
};

export class CreateMasterEdition extends Transaction {
  constructor(options: TransactionCtorFields, params: CreateMasterEditionParams) {
    super(options);
    const { feePayer } = options;
    const { edition, metadata, updateAuthority, mint, mintAuthority, maxSupply } = params;

    const data = CreateMasterEditionArgs.serialize({
      maxSupply: maxSupply || null,
    });

    this.add(
      new TransactionInstruction({
        keys: [
          {
            pubkey: edition,
            isSigner: false,
            isWritable: true,
          },
          {
            pubkey: mint,
            isSigner: false,
            isWritable: true,
          },
          {
            pubkey: updateAuthority,
            isSigner: true,
            isWritable: false,
          },
          {
            pubkey: mintAuthority,
            isSigner: true,
            isWritable: false,
          },
          {
            pubkey: feePayer,
            isSigner: true,
            isWritable: false,
          },
          {
            pubkey: metadata,
            isSigner: false,
            isWritable: false,
          },

          {
            pubkey: TOKEN_PROGRAM_ID,
            isSigner: false,
            isWritable: false,
          },
          {
            pubkey: SystemProgram.programId,
            isSigner: false,
            isWritable: false,
          },
          {
            pubkey: SYSVAR_RENT_PUBKEY,
            isSigner: false,
            isWritable: false,
          },
        ],
        programId: MetadataProgram.PUBKEY,
        data,
      }),
    );
  }
}

File: token-metadata/program/src/processor/deprecated/create_master_edition.rs
use borsh::BorshSerialize;
pub use instruction::*;
use solana_program::{
    instruction::{AccountMeta, Instruction},
    pubkey::Pubkey,
    sysvar,
};

use crate::{instruction::MetadataInstruction, processor::CreateMasterEditionArgs};

mod instruction {
    use super::*;

    /// creates a create_master_edition instruction
    #[allow(clippy::too_many_arguments)]
    /// [deprecated(since="1.1.0", note="please use `create_master_edition_v3` instead")]
    pub fn create_master_edition(
        program_id: Pubkey,
        edition: Pubkey,
        mint: Pubkey,
        update_authority: Pubkey,
        mint_authority: Pubkey,
        metadata: Pubkey,
        payer: Pubkey,
        max_supply: Option<u64>,
    ) -> Instruction {
        let accounts = vec![
            AccountMeta::new(edition, false),
            AccountMeta::new(mint, false),
            AccountMeta::new_readonly(update_authority, true),
            AccountMeta::new_readonly(mint_authority, true),
            AccountMeta::new(payer, true),
            AccountMeta::new_readonly(metadata, false),
            AccountMeta::new_readonly(spl_token::id(), false),
            AccountMeta::new_readonly(solana_program::system_program::id(), false),
            AccountMeta::new_readonly(sysvar::rent::id(), false),
        ];

        Instruction {
            program_id,
            accounts,
            data: MetadataInstruction::CreateMasterEdition(CreateMasterEditionArgs { max_supply })
                .try_to_vec()
                .unwrap(),
        }
    }
}
